File: haudiotagger/lib/src/pipeline.dart
import 'dart:typed_data';
import 'haudiotagger.dart';
import 'rust/api/api.dart' as api;
import 'rust/api/pipeline.dart' as rpc;

/// Result of previewing a pipeline transformation.
class PreviewResult {
  /// The original tag before transformation.
  final Tag original;

  /// The transformed tag after applying rules.
  final Tag transformed;

  /// The list of field changes.
  final List<FieldChange> changes;

  const PreviewResult({
    required this.original,
    required this.transformed,
    required this.changes,
  });

  /// Whether any changes were made.
  bool get hasChanges => changes.isNotEmpty;

  @override
  String toString() => 'PreviewResult(${changes.length} changes)';
}

/// A single field change between original and transformed tags.
class FieldChange {
  final String field;
  final String? oldValue;
  final String? newValue;

  const FieldChange({
    required this.field,
    this.oldValue,
    this.newValue,
  });

  @override
  String toString() => '$field: $oldValue → $newValue';
}

/// A pipeline of transformation rules to apply to audio metadata.
///
/// Example:
/// ```dart
/// final pipeline = TagPipeline()
///   ..trimWhitespace()
///   ..normalizeUnicode()
///   ..setAlbumArtist('Various Artists')
///   ..removeLyrics();
///
/// // Process multiple files
/// await pipeline.process(files);
///
/// // Preview changes on a single file
/// final result = await pipeline.preview('/path/to/song.mp3');
/// print(result.changes);
/// ```
class TagPipeline {
  final List<rpc.TransformRule> _rules;

  TagPipeline() : _rules = [];

  /// Create a pipeline from existing rules.
  factory TagPipeline.fromRules(List<rpc.TransformRule> rules) {
    return TagPipeline._(List.of(rules));
  }

  TagPipeline._(this._rules);

  // ── Whitespace / Unicode ──────────────────────────────

  /// Trim leading/trailing whitespace from all string fields.
  TagPipeline trimWhitespace() {
    _rules.add(const rpc.TransformRule.trimWhitespace());
    return this;
  }

  /// Collapse multiple whitespace characters into a single space.
  TagPipeline normalizeWhitespace() {
    _rules.add(const rpc.TransformRule.normalizeWhitespace());
    return this;
  }

  /// Apply Unicode NFKC normalization.
  TagPipeline normalizeUnicode() {
    _rules.add(const rpc.TransformRule.normalizeUnicode());
    return this;
  }

  // ── Setters ──────────────────────────────────────────

  /// Set the title.
  TagPipeline setTitle(String value) {
    _rules.add(rpc.TransformRule.setTitle(value));
    return this;
  }

  /// Set the artist.
  TagPipeline setArtist(String value) {
    _rules.add(rpc.TransformRule.setArtist(value));
    return this;
  }

  /// Set the album.
  TagPipeline setAlbum(String value) {
    _rules.add(rpc.TransformRule.setAlbum(value));
    return this;
  }

  /// Set the album artist.
  TagPipeline setAlbumArtist(String value) {
    _rules.add(rpc.TransformRule.setAlbumArtist(value));
    return this;
  }

  /// Set the genre.
  TagPipeline setGenre(String value) {
    _rules.add(rpc.TransformRule.setGenre(value));
    return this;
  }

  /// Set the year.
  TagPipeline setYear(int year) {
    _rules.add(rpc.TransformRule.setYear(year));
    return this;
  }

  /// Set the track number.
  TagPipeline setTrackNumber(int n) {
    _rules.add(rpc.TransformRule.setTrackNumber(n));
    return this;
  }

  /// Set the disc number.
  TagPipeline setDiscNumber(int n) {
    _rules.add(rpc.TransformRule.setDiscNumber(n));
    return this;
  }

  /// Set the track total.
  TagPipeline setTrackTotal(int n) {
    _rules.add(rpc.TransformRule.setTrackTotal(n));
    return this;
  }

  /// Set the disc total.
  TagPipeline setDiscTotal(int n) {
    _rules.add(rpc.TransformRule.setDiscTotal(n));
    return this;
  }

  /// Set the BPM.
  TagPipeline setBpm(double bpm) {
    _rules.add(rpc.TransformRule.setBpm(bpm));
    return this;
  }

  /// Set the comment.
  TagPipeline setComment(String value) {
    _rules.add(rpc.TransformRule.setComment(value));
    return this;
  }

  // ── Remove ───────────────────────────────────────────

  /// Remove lyrics.
  TagPipeline removeLyrics() {
    _rules.add(const rpc.TransformRule.removeLyrics());
    return this;
  }

  /// Remove comments.
  TagPipeline removeComment() {
    _rules.add(const rpc.TransformRule.removeComment());
    return this;
  }

  /// Remove all pictures.
  TagPipeline removePictures() {
    _rules.add(const rpc.TransformRule.removePictures());
    return this;
  }

  /// Remove BPM.
  TagPipeline removeBpm() {
    _rules.add(const rpc.TransformRule.removeBpm());
    return this;
  }

  /// Remove all ReplayGain fields.
  TagPipeline removeReplayGain() {
    _rules.add(const rpc.TransformRule.removeReplayGain());
    return this;
  }

  /// Remove title.
  TagPipeline removeTitle() {
    _rules.add(const rpc.TransformRule.removeTitle());
    return this;
  }

  /// Remove artist.
  TagPipeline removeArtist() {
    _rules.add(const rpc.TransformRule.removeArtist());
    return this;
  }

  /// Remove album.
  TagPipeline removeAlbum() {
    _rules.add(const rpc.TransformRule.removeAlbum());
    return this;
  }

  /// Remove album artist.
  TagPipeline removeAlbumArtist() {
    _rules.add(const rpc.TransformRule.removeAlbumArtist());
    return this;
  }

  /// Remove genre.
  TagPipeline removeGenre() {
    _rules.add(const rpc.TransformRule.removeGenre());
    return this;
  }

  /// Remove year.
  TagPipeline removeYear() {
    _rules.add(const rpc.TransformRule.removeYear());
    return this;
  }

  /// Remove track number.
  TagPipeline removeTrackNumber() {
    _rules.add(const rpc.TransformRule.removeTrackNumber());
    return this;
  }

  /// Remove disc number.
  TagPipeline removeDiscNumber() {
    _rules.add(const rpc.TransformRule.removeDiscNumber());
    return this;
  }

  // ── Normalize numbers ────────────────────────────────

  /// Normalize track numbers (clamp to valid range).
  TagPipeline normalizeTrackNumbers() {
    _rules.add(const rpc.TransformRule.normalizeTrackNumbers());
    return this;
  }

  /// Normalize disc numbers (clamp to valid range).
  TagPipeline normalizeDiscNumbers() {
    _rules.add(const rpc.TransformRule.normalizeDiscNumbers());
    return this;
  }

  /// Normalize year (2-digit to 4-digit).
  TagPipeline normalizeYear() {
    _rules.add(const rpc.TransformRule.normalizeYear());
    return this;
  }

  // ── Copy between fields ──────────────────────────────

  /// Copy artist to album artist (if album artist is empty).
  TagPipeline copyArtistToAlbumArtist() {
    _rules.add(const rpc.TransformRule.copyArtistToAlbumArtist());
    return this;
  }

  /// Copy album artist to artist (if artist is empty).
  TagPipeline copyAlbumArtistToArtist() {
    _rules.add(const rpc.TransformRule.copyAlbumArtistToArtist());
    return this;
  }

  /// Copy title to comment (if comment is empty).
  TagPipeline copyTitleToComment() {
    _rules.add(const rpc.TransformRule.copyTitleToComment());
    return this;
  }

  // ── Prefix / Suffix ──────────────────────────────────

  /// Add a prefix to the title.
  TagPipeline prefixTitle(String prefix) {
    _rules.add(rpc.TransformRule.prefixTitle(prefix));
    return this;
  }

  /// Add a suffix to the title.
  TagPipeline suffixTitle(String suffix) {
    _rules.add(rpc.TransformRule.suffixTitle(suffix));
    return this;
  }

  /// Add a prefix to the album.
  TagPipeline prefixAlbum(String prefix) {
    _rules.add(rpc.TransformRule.prefixAlbum(prefix));
    return this;
  }

  /// Add a suffix to the album.
  TagPipeline suffixAlbum(String suffix) {
    _rules.add(rpc.TransformRule.suffixAlbum(suffix));
    return this;
  }

  /// Add a prefix to the artist.
  TagPipeline prefixArtist(String prefix) {
    _rules.add(rpc.TransformRule.prefixArtist(prefix));
    return this;
  }

  /// Add a suffix to the artist.
  TagPipeline suffixArtist(String suffix) {
    _rules.add(rpc.TransformRule.suffixArtist(suffix));
    return this;
  }

  // ── Case transformations ─────────────────────────────

  /// Convert title to Title Case.
  TagPipeline titleCaseTitle() {
    _rules.add(const rpc.TransformRule.titleCaseTitle());
    return this;
  }

  /// Convert artist to Title Case.
  TagPipeline titleCaseArtist() {
    _rules.add(const rpc.TransformRule.titleCaseArtist());
    return this;
  }

  /// Convert album to Title Case.
  TagPipeline titleCaseAlbum() {
    _rules.add(const rpc.TransformRule.titleCaseAlbum());
    return this;
  }

  /// Convert all string fields to lowercase.
  TagPipeline lowerCaseAll() {
    _rules.add(const rpc.TransformRule.lowerCaseAll());
    return this;
  }

  /// Convert all string fields to uppercase.
  TagPipeline upperCaseAll() {
    _rules.add(const rpc.TransformRule.upperCaseAll());
    return this;
  }

  // ── Search / Replace ─────────────────────────────────

  /// Replace text in title.
  TagPipeline replaceInTitle(String find, String replace) {
    _rules.add(rpc.TransformRule.replaceInTitle(find: find, replace: replace));
    return this;
  }

  /// Replace text in artist.
  TagPipeline replaceInArtist(String find, String replace) {
    _rules.add(rpc.TransformRule.replaceInArtist(find: find, replace: replace));
    return this;
  }

  /// Replace text in album.
  TagPipeline replaceInAlbum(String find, String replace) {
    _rules.add(rpc.TransformRule.replaceInAlbum(find: find, replace: replace));
    return this;
  }

  /// Replace text in all string fields.
  TagPipeline replaceInAll(String find, String replace) {
    _rules.add(rpc.TransformRule.replaceInAll(find: find, replace: replace));
    return this;
  }

  // ── Conditional ──────────────────────────────────────

  /// Set title only if it is currently empty.
  TagPipeline setTitleIfEmpty(String value) {
    _rules.add(rpc.TransformRule.setTitleIfEmpty(value));
    return this;
  }

  /// Set artist only if it is currently empty.
  TagPipeline setArtistIfEmpty(String value) {
    _rules.add(rpc.TransformRule.setArtistIfEmpty(value));
    return this;
  }

  /// Set album only if it is currently empty.
  TagPipeline setAlbumIfEmpty(String value) {
    _rules.add(rpc.TransformRule.setAlbumIfEmpty(value));
    return this;
  }

  /// Set genre only if it is currently empty.
  TagPipeline setGenreIfEmpty(String value) {
    _rules.add(rpc.TransformRule.setGenreIfEmpty(value));
    return this;
  }

  /// Set album artist only if it is currently empty.
  TagPipeline setAlbumArtistIfEmpty(String value) {
    _rules.add(rpc.TransformRule.setAlbumArtistIfEmpty(value));
    return this;
  }

  // ── Remove empty ─────────────────────────────────────

  /// Remove all fields that are empty strings or null.
  TagPipeline removeEmptyFields() {
    _rules.add(const rpc.TransformRule.removeEmptyFields());
    return this;
  }

  // ── Pictures ─────────────────────────────────────────

  /// Remove all pictures except cover art.
  TagPipeline removeNonCoverPictures() {
    _rules.add(const rpc.TransformRule.removeNonCoverPictures());
    return this;
  }

  // ── Execution ────────────────────────────────────────

  /// Apply all rules to a tag and return the transformed tag.
  Future<Tag> apply(Tag tag) async {
    final pipeline = await rpc.TagPipeline.fromRules(rules: _rules);
    return pipeline.apply(tag: tag);
  }

  /// Preview the changes that would be made to a file.
  ///
  /// Returns a [PreviewResult] with the original tag, transformed tag,
  /// and a list of [FieldChange]s.
  Future<PreviewResult> preview(String path) async {
    final original = await Haudiotagger.read(path);
    if (original == null) {
      throw Exception('No metadata found at $path');
    }
    final transformed = await apply(original);
    final changes = _diffTags(original, transformed);
    return PreviewResult(
      original: original,
      transformed: transformed,
      changes: changes,
    );
  }

  /// Preview the changes that would be made to a byte array.
  Future<PreviewResult> previewFromBytes(Uint8List bytes) async {
    final original = await Haudiotagger.readFromBytes(bytes);
    if (original == null) {
      throw Exception('No metadata found in bytes');
    }
    final transformed = await apply(original);
    final changes = _diffTags(original, transformed);
    return PreviewResult(
      original: original,
      transformed: transformed,
      changes: changes,
    );
  }

  /// Apply the pipeline to a single file and write the result.
  Future<void> processFile(String path) async {
    await api.processFile(path: path, rules: _rules);
  }

  /// Apply the pipeline to a byte array and return the modified bytes.
  Future<Uint8List> processFromBytes(Uint8List bytes) async {
    final result = await api.processBytes(bytes: bytes, rules: _rules);
    return result;
  }

  /// Apply the pipeline to multiple files.
  ///
  /// Returns the number of files successfully processed.
  Future<int> process(List<String> files) async {
    final result = await api.processBatch(paths: files, rules: _rules);
    return result.successes;
  }

  /// Apply the pipeline to multiple byte arrays.
  ///
  /// Returns the list of modified byte arrays (same length as input).
  Future<List<Uint8List>> processBatchFromBytes(
    List<Uint8List> byteArrays,
  ) async {
    final result = await api.processBatchBytes(
      byteArrays: byteArrays,
      rules: _rules,
    );
    return result.results.toList();
  }

  /// Get the list of rules in this pipeline.
  List<rpc.TransformRule> get rules => List.unmodifiable(_rules);

  /// Get the number of rules in this pipeline.
  int get length => _rules.length;

  /// Check if the pipeline has no rules.
  bool get isEmpty => _rules.isEmpty;

  List<FieldChange> _diffTags(Tag a, Tag b) {
    final changes = <FieldChange>[];

    void check(String name, String? av, String? bv) {
      if (av != bv) {
        changes.add(FieldChange(field: name, oldValue: av, newValue: bv));
      }
    }

    check('title', a.title, b.title);
    check('artist', a.trackArtist, b.trackArtist);
    check('album', a.album, b.album);
    check('albumArtist', a.albumArtist, b.albumArtist);
    check('genre', a.genre, b.genre);
    check('lyrics', a.lyrics, b.lyrics);
    check('comment', a.comment, b.comment);
    check('year', a.year?.toString(), b.year?.toString());
    check('trackNumber', a.trackNumber?.toString(), b.trackNumber?.toString());
    check('trackTotal', a.trackTotal?.toString(), b.trackTotal?.toString());
    check('discNumber', a.discNumber?.toString(), b.discNumber?.toString());
    check('discTotal', a.discTotal?.toString(), b.discTotal?.toString());
    check('bpm', a.bpm?.toString(), b.bpm?.toString());

    if (a.pictures.length != b.pictures.length) {
      changes.add(FieldChange(
        field: 'pictures',
        oldValue: '${a.pictures.length} pictures',
        newValue: '${b.pictures.length} pictures',
      ));
    }

    return changes;
  }
}

File: haudiotagger/rust/src/api/pipeline.rs
use unicode_normalization::UnicodeNormalization;

use super::tag::Tag;

/// A single transformation rule that can be applied to a `Tag`.
#[derive(Debug, Clone)]
pub enum TransformRule {
    // ── Whitespace / Unicode ──────────────────────────────
    TrimWhitespace,
    NormalizeWhitespace,
    NormalizeUnicode,

    // ── Setters ──────────────────────────────────────────
    SetTitle(String),
    SetArtist(String),
    SetAlbum(String),
    SetAlbumArtist(String),
    SetGenre(String),
    SetYear(u32),
    SetTrackNumber(u32),
    SetDiscNumber(u32),
    SetTrackTotal(u32),
    SetDiscTotal(u32),
    SetBpm(f32),
    SetComment(String),

    // ── Remove ───────────────────────────────────────────
    RemoveLyrics,
    RemoveComment,
    RemovePictures,
    RemoveBpm,
    RemoveReplayGain,
    RemoveTitle,
    RemoveArtist,
    RemoveAlbum,
    RemoveAlbumArtist,
    RemoveGenre,
    RemoveYear,
    RemoveTrackNumber,
    RemoveDiscNumber,

    // ── Normalize numbers ────────────────────────────────
    NormalizeTrackNumbers,
    NormalizeDiscNumbers,
    NormalizeYear,

    // ── Copy between fields ──────────────────────────────
    CopyArtistToAlbumArtist,
    CopyAlbumArtistToArtist,
    CopyTitleToComment,

    // ── Prefix / Suffix ──────────────────────────────────
    PrefixTitle(String),
    SuffixTitle(String),
    PrefixAlbum(String),
    SuffixAlbum(String),
    PrefixArtist(String),
    SuffixArtist(String),

    // ── Case transformations ─────────────────────────────
    TitleCaseTitle,
    TitleCaseArtist,
    TitleCaseAlbum,
    LowerCaseAll,
    UpperCaseAll,

    // ── Search / Replace ─────────────────────────────────
    ReplaceInTitle { find: String, replace: String },
    ReplaceInArtist { find: String, replace: String },
    ReplaceInAlbum { find: String, replace: String },
    ReplaceInAll { find: String, replace: String },

    // ── Conditional ──────────────────────────────────────
    SetTitleIfEmpty(String),
    SetArtistIfEmpty(String),
    SetAlbumIfEmpty(String),
    SetGenreIfEmpty(String),
    SetAlbumArtistIfEmpty(String),

    // ── Remove empty ─────────────────────────────────────
    RemoveEmptyFields,

    // ── Pictures ─────────────────────────────────────────
    RemoveNonCoverPictures,
}

/// A pipeline of transformation rules to apply to tags.
#[derive(Debug, Clone, Default)]
pub struct TagPipeline {
    pub rules: Vec<TransformRule>,
}

impl TagPipeline {
    /// Create a new empty pipeline.
    pub fn new() -> Self {
        Self::default()
    }

    /// Create a pipeline from a list of rules.
    pub fn from_rules(rules: Vec<TransformRule>) -> Self {
        Self { rules }
    }

    /// Apply all rules to a tag and return the transformed tag.
    pub fn apply(&self, tag: &Tag) -> Tag {
        let mut result = tag.clone();
        for rule in &self.rules {
            apply_rule(&mut result, rule);
        }
        result
    }

    /// Get the number of rules in this pipeline.
    pub fn len(&self) -> usize {
        self.rules.len()
    }

    /// Check if the pipeline has no rules.
    pub fn is_empty(&self) -> bool {
        self.rules.is_empty()
    }
}

// ── Helpers ───────────────────────────────────────────────

fn trim(s: &str) -> String {
    s.trim().to_string()
}

fn collapse_ws(s: &str) -> String {
    let mut out = String::new();
    for (i, word) in s.split_whitespace().enumerate() {
        if i > 0 {
            out.push(' ');
        }
        out.push_str(word);
    }
    out
}

fn nfkc(s: &str) -> String {
    s.nfkc().collect()
}

fn title_case(s: &str) -> String {
    s.split_whitespace()
        .map(|word| {
            let mut chars = word.chars();
            match chars.next() {
                None => String::new(),
                Some(first) => {
                    let upper: String = first.to_uppercase().collect();
                    let rest: String = chars.collect::<String>().to_lowercase();
                    format!("{upper}{rest}")
                }
            }
        })
        .collect::<Vec<String>>()
        .join(" ")
}

fn normalize_year(year: u32) -> u32 {
    if year > 0 && year < 100 {
        if year >= 30 { 1900 + year } else { 2000 + year }
    } else {
        year
    }
}

fn apply_to_strings(tag: &mut Tag, f: impl Fn(&str) -> String) {
    tag.title = tag.title.as_ref().map(|s| f(s));
    tag.track_artist = tag.track_artist.as_ref().map(|s| f(s));
    tag.album = tag.album.as_ref().map(|s| f(s));
    tag.album_artist = tag.album_artist.as_ref().map(|s| f(s));
    tag.genre = tag.genre.as_ref().map(|s| f(s));
    tag.lyrics = tag.lyrics.as_ref().map(|s| f(s));
    tag.comment = tag.comment.as_ref().map(|s| f(s));
    tag.replay_gain_track_gain = tag.replay_gain_track_gain.as_ref().map(|s| f(s));
    tag.replay_gain_track_peak = tag.replay_gain_track_peak.as_ref().map(|s| f(s));
    tag.replay_gain_album_gain = tag.replay_gain_album_gain.as_ref().map(|s| f(s));
    tag.replay_gain_album_peak = tag.replay_gain_album_peak.as_ref().map(|s| f(s));
}

// ── Rule application ──────────────────────────────────────

/// Apply a single rule to a tag.
pub fn apply_rule(tag: &mut Tag, rule: &TransformRule) {
    match rule {
        // ── Whitespace / Unicode ─────────────────────────
        TransformRule::TrimWhitespace => apply_to_strings(tag, trim),
        TransformRule::NormalizeWhitespace => apply_to_strings(tag, collapse_ws),
        TransformRule::NormalizeUnicode => apply_to_strings(tag, nfkc),

        // ── Setters ─────────────────────────────────────
        TransformRule::SetTitle(v) => tag.title = Some(v.clone()),
        TransformRule::SetArtist(v) => tag.track_artist = Some(v.clone()),
        TransformRule::SetAlbum(v) => tag.album = Some(v.clone()),
        TransformRule::SetAlbumArtist(v) => tag.album_artist = Some(v.clone()),
        TransformRule::SetGenre(v) => tag.genre = Some(v.clone()),
        TransformRule::SetYear(y) => tag.year = Some(*y),
        TransformRule::SetTrackNumber(n) => tag.track_number = Some(*n),
        TransformRule::SetDiscNumber(n) => tag.disc_number = Some(*n),
        TransformRule::SetTrackTotal(n) => tag.track_total = Some(*n),
        TransformRule::SetDiscTotal(n) => tag.disc_total = Some(*n),
        TransformRule::SetBpm(b) => tag.bpm = Some(*b),
        TransformRule::SetComment(v) => tag.comment = Some(v.clone()),

        // ── Remove ──────────────────────────────────────
        TransformRule::RemoveLyrics => tag.lyrics = None,
        TransformRule::RemoveComment => tag.comment = None,
        TransformRule::RemovePictures => tag.pictures = vec![],
        TransformRule::RemoveBpm => tag.bpm = None,
        TransformRule::RemoveReplayGain => {
            tag.replay_gain_track_gain = None;
            tag.replay_gain_track_peak = None;
            tag.replay_gain_album_gain = None;
            tag.replay_gain_album_peak = None;
        }
        TransformRule::RemoveTitle => tag.title = None,
        TransformRule::RemoveArtist => tag.track_artist = None,
        TransformRule::RemoveAlbum => tag.album = None,
        TransformRule::RemoveAlbumArtist => tag.album_artist = None,
        TransformRule::RemoveGenre => tag.genre = None,
        TransformRule::RemoveYear => tag.year = None,
        TransformRule::RemoveTrackNumber => tag.track_number = None,
        TransformRule::RemoveDiscNumber => tag.disc_number = None,

        // ── Normalize numbers ───────────────────────────
        TransformRule::NormalizeTrackNumbers => {
            tag.track_number = tag.track_number.map(|n| n.clamp(0, 999));
        }
        TransformRule::NormalizeDiscNumbers => {
            tag.disc_number = tag.disc_number.map(|n| n.clamp(0, 99));
        }
        TransformRule::NormalizeYear => {
            tag.year = tag.year.map(normalize_year);
        }

        // ── Copy between fields ─────────────────────────
        TransformRule::CopyArtistToAlbumArtist => {
            if tag.album_artist.as_deref() == Some("") || tag.album_artist.is_none() {
                tag.album_artist = tag.track_artist.clone();
            }
        }
        TransformRule::CopyAlbumArtistToArtist => {
            if tag.track_artist.as_deref() == Some("") || tag.track_artist.is_none() {
                tag.track_artist = tag.album_artist.clone();
            }
        }
        TransformRule::CopyTitleToComment => {
            if tag.comment.as_deref() == Some("") || tag.comment.is_none() {
                tag.comment = tag.title.clone();
            }
        }

        // ── Prefix / Suffix ─────────────────────────────
        TransformRule::PrefixTitle(p) => {
            tag.title = tag.title.as_ref().map(|s| format!("{p}{s}"));
        }
        TransformRule::SuffixTitle(s) => {
            tag.title = tag.title.as_ref().map(|t| format!("{t}{s}"));
        }
        TransformRule::PrefixAlbum(p) => {
            tag.album = tag.album.as_ref().map(|a| format!("{p}{a}"));
        }
        TransformRule::SuffixAlbum(s) => {
            tag.album = tag.album.as_ref().map(|a| format!("{a}{s}"));
        }
        TransformRule::PrefixArtist(p) => {
            tag.track_artist = tag.track_artist.as_ref().map(|a| format!("{p}{a}"));
        }
        TransformRule::SuffixArtist(s) => {
            tag.track_artist = tag.track_artist.as_ref().map(|a| format!("{a}{s}"));
        }

        // ── Case transformations ────────────────────────
        TransformRule::TitleCaseTitle => {
            tag.title = tag.title.as_ref().map(|s| title_case(s));
        }
        TransformRule::TitleCaseArtist => {
            tag.track_artist = tag.track_artist.as_ref().map(|s| title_case(s));
        }
        TransformRule::TitleCaseAlbum => {
            tag.album = tag.album.as_ref().map(|s| title_case(s));
        }
        TransformRule::LowerCaseAll => apply_to_strings(tag, |s| s.to_lowercase()),
        TransformRule::UpperCaseAll => apply_to_strings(tag, |s| s.to_uppercase()),

        // ── Search / Replace ────────────────────────────
        TransformRule::ReplaceInTitle { find, replace } => {
            tag.title = tag.title.as_ref().map(|s| s.replace(find, replace));
        }
        TransformRule::ReplaceInArtist { find, replace } => {
            tag.track_artist = tag.track_artist.as_ref().map(|s| s.replace(find, replace));
        }
        TransformRule::ReplaceInAlbum { find, replace } => {
            tag.album = tag.album.as_ref().map(|s| s.replace(find, replace));
        }
        TransformRule::ReplaceInAll { find, replace } => {
            apply_to_strings(tag, |s| s.replace(find, replace));
        }

        // ── Conditional ─────────────────────────────────
        TransformRule::SetTitleIfEmpty(v) => {
            if tag.title.as_deref() == Some("") || tag.title.is_none() {
                tag.title = Some(v.clone());
            }
        }
        TransformRule::SetArtistIfEmpty(v) => {
            if tag.track_artist.as_deref() == Some("") || tag.track_artist.is_none() {
                tag.track_artist = Some(v.clone());
            }
        }
        TransformRule::SetAlbumIfEmpty(v) => {
            if tag.album.as_deref() == Some("") || tag.album.is_none() {
                tag.album = Some(v.clone());
            }
        }
        TransformRule::SetGenreIfEmpty(v) => {
            if tag.genre.as_deref() == Some("") || tag.genre.is_none() {
                tag.genre = Some(v.clone());
            }
        }
        TransformRule::SetAlbumArtistIfEmpty(v) => {
            if tag.album_artist.as_deref() == Some("") || tag.album_artist.is_none() {
                tag.album_artist = Some(v.clone());
            }
        }

        // ── Remove empty ────────────────────────────────
        TransformRule::RemoveEmptyFields => {
            tag.title = tag.title.as_ref().and_then(|s| {
                let t = s.trim();
                if t.is_empty() {
                    None
                } else {
                    Some(t.to_string())
                }
            });
            tag.track_artist = tag.track_artist.as_ref().and_then(|s| {
                let t = s.trim();
                if t.is_empty() {
                    None
                } else {
                    Some(t.to_string())
                }
            });
            tag.album = tag.album.as_ref().and_then(|s| {
                let t = s.trim();
                if t.is_empty() {
                    None
                } else {
                    Some(t.to_string())
                }
            });
            tag.album_artist = tag.album_artist.as_ref().and_then(|s| {
                let t = s.trim();
                if t.is_empty() {
                    None
                } else {
                    Some(t.to_string())
                }
            });
            tag.genre = tag.genre.as_ref().and_then(|s| {
                let t = s.trim();
                if t.is_empty() {
                    None
                } else {
                    Some(t.to_string())
                }
            });
            tag.lyrics = tag.lyrics.as_ref().and_then(|s| {
                let t = s.trim();
                if t.is_empty() {
                    None
                } else {
                    Some(t.to_string())
                }
            });
            tag.comment = tag.comment.as_ref().and_then(|s| {
                let t = s.trim();
                if t.is_empty() {
                    None
                } else {
                    Some(t.to_string())
                }
            });
        }

        // ── Pictures ────────────────────────────────────
        TransformRule::RemoveNonCoverPictures => {
            tag.pictures.retain(|p| {
                matches!(
                    p.picture_type,
                    super::picture::PictureType::CoverFront
                        | super::picture::PictureType::CoverBack
                )
            });
        }
    }
}

/// Apply a pipeline of rules to a tag.
pub fn apply_pipeline(tag: &Tag, pipeline: &TagPipeline) -> Tag {
    pipeline.apply(tag)
}
